src/torchcodec/samplers/_implem.py

```diff
@@ -323,3 +331,236 @@ def clips_at_regular_indices(
+        # We allow a clip to start anywhere within
+        # [sampling_range_start, sampling_range_end)
+        # When sampling_range_end is None, we want to automatically set it to
+        # the largest possible value such that the sampled frames in any clip
+        # are within the bounds of the video duration (in other words, we don't
+        # want to have to resort to the `policy`).
+        # I.e. we want to guarantee that for all frames in any clip we have
+        # pts < end_stream_seconds.
+        #
+        # The frames of a clip will be sampled at the following pts:
+        # clip_timestamps = [
+        #  clip_start + 0 * seconds_between_frames,
+        #  clip_start + 1 * seconds_between_frames,
+        #  clip_start + 2 * seconds_between_frames,
+        #  ...
+        #  clip_start + (num_frames_per_clip - 1) * seconds_between_frames,
+        # ]
+        # To guarantee that any such value is < end_stream_seconds, we only need
+        # to guarantee that
+        # clip_start < end_stream_seconds - (num_frames_per_clip - 1) * seconds_between_frames
+        #
+        # So that's the value of sampling_range_end we want to use.
+        sampling_range_end = (
+            end_stream_seconds - (num_frames_per_clip - 1) * seconds_between_frames
+        )
+    elif sampling_range_end <= begin_stream_seconds:
+        raise ValueError(
+            f"sampling_range_end ({sampling_range_end}) must be at least {begin_stream_seconds}"
+        )
+
+    if sampling_range_start >= sampling_range_end:
+        raise ValueError(
+            f"sampling_range_start ({sampling_range_start}) must be smaller than sampling_range_end ({sampling_range_end})"
+        )
+
+    sampling_range_end = min(sampling_range_end, end_stream_seconds)
+
+    return sampling_range_start, sampling_range_end
+
+
+def _build_all_clips_timestamps(
+    *,
+    clip_start_seconds: torch.Tensor,  # 1D float tensor
+    num_frames_per_clip: int,
+    seconds_between_frames: float,
+    end_stream_seconds: float,
+    policy_fun: _POLICY_FUNCTION_TYPE,
+) -> list[float]:
+
+    all_clips_timestamps: list[float] = []
+    for start_seconds in clip_start_seconds:
+        clip_timestamps = [
+            start_seconds + i * seconds_between_frames
+            for i in range(num_frames_per_clip)
+        ]
+        clip_timestamps = [
+            timestamp for timestamp in clip_timestamps if timestamp < end_stream_seconds
+        ]
+
+        if len(clip_timestamps) < num_frames_per_clip:
+            clip_timestamps = policy_fun(clip_timestamps, num_frames_per_clip)
+        all_clips_timestamps += clip_timestamps
+
+    return all_clips_timestamps
+
+
+def _decode_all_clips_timestamps(
+    decoder: VideoDecoder, all_clips_timestamps: list[float], num_frames_per_clip: int
+) -> list[FrameBatch]:
+    # This is 99% the same as _decode_all_clips_indices. The only change is the
+    # call to .get_frame_displayed_at(pts) instead of .get_frame_at(idx)
```

Comment:
Could we at least extract out `chunk_list()` and `to_framebatch()` as reusable components?

Reply:
We could also pass in the method to call for the actual decoding. But we can work on that later.

Reply:
Done.

> We could also pass in the method to call for the actual decoding. But we can work on that later.

I thought about doing this. We can factorize this code to death, even the input validation and the `_build_all_clips_indices` / `_build_all_clips_timestamps` could be made generic.

For `_decode_all_clips_timestamps` / `_decode_all_clips_indices`, I feel like it's not worth it because we'll be re-writing them in C++ soon anyway. For the rest, let's leave it out for later. There will be a trade-off between compactness and clarity of the code.